torchci/pages/hud/[repoOwner]/[repoName]/[branch]/[[...page]].tsx

```diff
@@ -452,6 +466,82 @@ export default function Hud() {
   );
 }
 
+function useLatestCommitSha(params: HudParams) {
+  const data = useHudData(params);
+  if (data === undefined) {
+    return null;
+  }
+  if (data.shaGrid.length === 0) {
+    return null; // Nothing worth copying
+  }
+  if (data.shaGrid[0].sha === undefined) {
+    return null; // No sha to copy. This should never happen (TM)
+  }
+
+  return data.shaGrid[0].sha;
+}
+
+function usePermalinkCopiedState(params: HudParams) {
+  // We ensure that the "Copied" state is reset when the params change
+  // since that would change the permalink url.
+  // Params can change on things like page navigation, changing job filter, etc.
+
+  const [copied, setCopied] = useState(false);
+
+  // Store the old params to know when the actual params
+  // change instead of just the reference to them
+  const prevParamsRef = useRef(params);
+
+  useEffect(() => {
+    if (!_.isEqual(prevParamsRef.current, params)) {
+      prevParamsRef.current = params;
+      setCopied(false);
+    }
+  }, [params]);
+
+  return { copied, setCopied };
+}
+
+function CopyPermanentLink({
+  params,
+  style,
+}: {
+  params: HudParams;
+  style?: React.CSSProperties;
+}) {
+  // Used to let users know that the permalink has been successfully copied to their clipboard
+  const { copied, setCopied } = usePermalinkCopiedState(params);
+
+  // Branch and tag pointers can change over time.
+  // For a permanent, we take the latest immutable commit as our reference
+  const latestCommitSha = useLatestCommitSha(params);
+  if (latestCommitSha === null) {
+    return <></>;
+  }
+  let permaParams = { ...params, branch: latestCommitSha };
+
+  const domain = window.location.origin;
+  const path = formatHudUrlForRoute("hud", permaParams);
+  const url = `${domain}${path}`;
+  return (
+    <>
+      <button
+        onClick={() => {
+          navigator.clipboard.writeText(url);
+          setCopied(true);
+
+          setTimeout(() => {
+            setCopied(false);
+          }, 10 * 1000); // 10000 milliseconds = 10 seconds
+        }}
+        style={style}
+      >
+        {copied ? "📋 ✅ Copied" : "📋 Permalink"}
```

Comment:
```suggestion
        {copied ? "✅ Copied" : "🔗 Permalink"}
```

Just a nit, I think making these icons more clear will make it easier to understand

Reply:
I just realize another convention we are using for this is in minihud page https://hud.pytorch.org/minihud (Copy Link).  Click on that and the a different check mark style shows up indicating that the link has been copied.  Two thoughts:

* We should decide on one style, and make `CopyPermanentLink` reusable
* The minuhud link is broken, i.e. https://hud.pytorch.org/minihud#980f5ac0499447cd6d39b6241ae30ae30937a3e5 (not related to this change though)

Reply:
I like it

Reply:
Weird, now it doesn't seem to crash anymore

Reply:
Re: choosing a convention, let's stick with the 🔗, which is more standard (and [Google agrees](https://www.google.com/search?q=usual+emoji++used+for+permalink&sca_esv=33d761b251b7bc29&sca_upv=1&ei=tvdMZoWlBsDCp84P36a66A4&ved=0ahUKEwiF2Jn2vp-GAxVA4ckDHV-TDu0Q4dUDCBA&uact=5&oq=usual+emoji++used+for+permalink&gs_lp=Egxnd3Mtd2l6LXNlcnAiH3VzdWFsIGVtb2ppICB1c2VkIGZvciBwZXJtYWxpbmsyCBAhGKABGMMEMggQIRigARjDBEiCF1CvC1ipEnADeAGQAQCYAWegAaEEqgEDNS4xuAEDyAEA-AEBmAIJoAKxBMICChAAGLADGNYEGEfCAgoQIRigARjDBBgKmAMAiAYBkAYIkgcDOC4xoAfFGQ&sclient=gws-wiz-serp))

I've updated the icons of minihud to also be the same. I can look into fixing the bug with minihud's copy link as a separate issue.